src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs

```diff
@@ -5635,6 +5661,146 @@ public void Utf8Output()
+}} | {s_CSharpCompilerExecutable} /nologo /t:exe -""", workingDirectory: tempDir,
+                    // we are testing shell's piped/redirected stdin behavior explicitly
+                    // instead of using Process.StandardInput.Write(), so we set
+                    // redirectStandardInput to true, which implies that isatty of child
+                    // process is false and thereby Console.IsInputRedirected will return
+                    // true in csc code.
+                    redirectStandardInput: true);
+
+            Assert.False(result.ContainsErrors, $"Compilation error(s) occurred: {result.Output} {result.Errors}");
+
+            string output = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ?
+                ProcessUtilities.RunAndGetOutput("cmd.exe", $@"/C ""{s_DotnetCscRun} -.exe""", expectedRetCode: 0, startFolder: tempDir) :
+                ProcessUtilities.RunAndGetOutput("sh", $@"-c ""{s_DotnetCscRun} -.exe""", expectedRetCode: 0, startFolder: tempDir);
+
+            Assert.Equal("Hello World!", output.Trim());
+        }
+
+        [Fact]
+        public void CscCompile_WithSourceCodeRedirectedViaStandardInput_ProducesLibrary()
+        {
+            var name = Guid.NewGuid().ToString() + ".dll";
+            string tempDir = Temp.CreateDirectory().Path;
+            ProcessResult result = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ?
+                ProcessUtilities.Run("cmd", $@"/C echo  ^
+class A                                                 ^
+{{                                                      ^
+    public A Get() =^^^> default;                       ^
+}} | {s_CSharpCompilerExecutable} /nologo /t:library /out:{name} -"
+    .Replace(Environment.NewLine, string.Empty), workingDirectory: tempDir) :
+                ProcessUtilities.Run("/usr/bin/env", $@"sh -c ""echo  \
+class A                                                               \
+{{                                                                    \
+    public A Get\(\) =\> default\;                                    \
+}} | {s_CSharpCompilerExecutable} /nologo /t:library /out:{name} -""", workingDirectory: tempDir,
+                    // we are testing shell's piped/redirected stdin behavior explicitly
+                    // instead of using Process.StandardInput.Write(), so we set
+                    // redirectStandardInput to true, which implies that isatty of child
+                    // process is false and thereby Console.IsInputRedirected will return
+                    // true in csc code.
+                    redirectStandardInput: true);
+
+            Assert.False(result.ContainsErrors, $"Compilation error(s) occurred: {result.Output} {result.Errors}");
+
+            var assembly = Assembly.LoadFrom(Path.Combine(tempDir, name));
+            assembly.ManifestModule.GetPEKind(out PortableExecutableKinds peKind, out _);
+
+            Assert.Equal(PortableExecutableKinds.ILOnly, peKind);
```

Comment:
We should avoid using `Assembly.LoadFrom()` in tests since that will pollute the assembly space. 

If the goal is to test that a valid assembly has been generated, would it be sufficient to use `CLRHelpers.PeVerify(string path)` instead?

Reply:
I haven't used CLRHelpers.PeVerify before, but it looks like net472 only. Is there a separate port for netcoreapp2?

Reply:
What is the core goal here though for `Assembly.LoadFrom`? If the goal is to inspect that the assembly is valid in some way generally we can use alternate loads that don't pollute the assembly space to verify the contents.

---
In reply to: [374409125](https://github.com/dotnet/roslyn/pull/41166#discussion_r374409125) [](ancestors = 374409125)

Reply:
> inspect that the assembly is valid in some way generally

Correct, to test the validity of produced assembly. I have changed it to `AssemblyName.GetAssemblyName(path)` which throws `BadImageFormatException` when assembly at `path` is invalid.

Reply:
I think this may work https://github.com/dotnet/roslyn/blob/2ed60699f9961a451f60d560da8bd4788c0182fc/src/Test/Utilities/Portable/Compilation/IRuntimeEnvironment.cs#L28

If you can construct a runtime environment from the emitted module, PEVerify should "just work"

Reply:
These are still unimplemented: https://github.com/dotnet/roslyn/blob/2ed60699f9961a451f60d560da8bd4788c0182fc/src/Test/Utilities/Portable/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs#L104-L115

so test will be a no-op on netcoreapp3.1.